src/services/machines/siopV2MachineService.ts

```diff
@@ -0,0 +1,158 @@
+  const {requestData} = context;
+
+  if (requestData?.uri === undefined) {
+    return Promise.reject(Error('Missing request uri in context'));
+  }
+
+  const identifier: IIdentifier = await getOrCreatePrimaryIdentifier();
+  return {
+    id: uuidv4(),
+    // FIXME: Update these values in SSI-SDK. Only the URI (not a redirectURI) would be available at this point
+    sessionId: uuidv4(),
+    redirectUrl: requestData.uri,
+    stateId: requestData.state,
+    identifier,
+  };
+};
+
+export const getSiopRequest = async (
+  context: Pick<SiopV2MachineContext, 'didAuthConfig' | 'requestData'>,
+): Promise<SiopV2AuthorizationRequestData> => {
+  const {didAuthConfig, requestData} = context;
+
+  if (requestData?.uri === undefined) {
+    return Promise.reject(Error('Missing request uri in context'));
+  }
+
+  if (didAuthConfig === undefined) {
+    return Promise.reject(Error('Missing config in context'));
+  }
+
+  const verifiedAuthorizationRequest: VerifiedAuthorizationRequest = await siopGetRequest(didAuthConfig);
+  const name = verifiedAuthorizationRequest.registrationMetadataPayload?.client_name;
+  const url =
+    verifiedAuthorizationRequest.responseURI ??
+    (requestData.uri.includes('request_uri')
+      ? decodeURIComponent(requestData.uri.split('?request_uri=')[1].trim())
+      : verifiedAuthorizationRequest.issuer ?? verifiedAuthorizationRequest.registrationMetadataPayload?.client_id);
+  const uri: URL | undefined = url.includes('://') ? new URL(url) : undefined;
+  const correlationIdName = uri
+    ? translateCorrelationIdToName(uri.hostname)
+    : verifiedAuthorizationRequest.issuer
+    ? translateCorrelationIdToName(verifiedAuthorizationRequest.issuer.split('://')[1])
+    : name;
+  const correlationId: string = uri?.hostname ?? correlationIdName;
+  const clientId: string | undefined = await verifiedAuthorizationRequest.authorizationRequest.getMergedProperty<string>('client_id');
+
+  return {
+    issuer: verifiedAuthorizationRequest.issuer,
+    correlationId,
+    registrationMetadataPayload: verifiedAuthorizationRequest.registrationMetadataPayload,
+    uri,
+    name,
+    clientId,
+    presentationDefinitions: verifiedAuthorizationRequest.presentationDefinitions,
+  };
+};
+
+export const retrieveContact = async (
+  context: Pick<SiopV2MachineContext, 'requestData' | 'authorizationRequestData'>,
+): Promise<IContact | undefined> => {
+  const {authorizationRequestData} = context;
+
+  if (authorizationRequestData === undefined) {
+    return Promise.reject(Error('Missing authorization request data in context'));
+  }
+
+  return getContacts({
+    filter: [
+      {
+        identities: {
+          identifier: {
+            correlationId: authorizationRequestData.correlationId,
+          },
+        },
+      },
+    ],
+  }).then((contacts: Array<IContact>): IContact | undefined => (contacts.length === 1 ? contacts[0] : undefined));
```

Comment:
contacts?.length === 1 ?

Reply:
contacts is never undefined but always an array with 1 or more items